src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/ExtractToCodeBehindCodeActionResolver.cs

```diff
@@ -0,0 +1,184 @@
+            var path = actionParams.Uri.GetAbsoluteOrUNCPath();
+
+            var document = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(path, out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler);
+            if (document is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await document.GetGeneratedOutputAsync();
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var codeDocumentFileKind = codeDocument.GetFileKind();
+            if (!FileKinds.IsComponent(codeDocumentFileKind))
+            {
+                return null;
+            }
+
+            var codeBehindPath = GenerateCodeBehindPath(path);
+            var codeBehindUri = new UriBuilder()
+            {
+                Scheme = Uri.UriSchemeFile,
+                Path = codeBehindPath,
+                Host = string.Empty,
+            }.Uri;
+
+            var text = await document.GetTextAsync();
+            if (text is null)
+            {
+                return null;
+            }
+
+            var className = Path.GetFileNameWithoutExtension(path);
+            var codeBlockcontent = text.GetSubTextString(new CodeAnalysis.Text.TextSpan(actionParams.ExtractStart, actionParams.ExtractEnd - actionParams.ExtractStart));
+            var codeBehindContent = GenerateCodeBehindClass(className, codeBlockcontent, codeDocument);
+
+            var start = codeDocument.Source.Lines.GetLocation(actionParams.RemoveStart);
+            var end = codeDocument.Source.Lines.GetLocation(actionParams.RemoveEnd);
+            var removeRange = new Range(
+                new Position(start.LineIndex, start.CharacterIndex),
+                new Position(end.LineIndex, end.CharacterIndex));
+
+            var changes = new Dictionary<Uri, IEnumerable<TextEdit>>
+            {
+                [actionParams.Uri] = new[]
+                {
+                    new TextEdit()
+                    {
+                        NewText = "",
+                        Range = removeRange,
+                    }
+                },
+                [codeBehindUri] = new[]
+                {
+                    new TextEdit()
+                    {
+                        NewText = codeBehindContent,
+                        Range = StartOfDocumentRange,
+                    }
+                }
+            };
+
+            var documentChanges = new List<WorkspaceEditDocumentChange>
+            {
+                new WorkspaceEditDocumentChange(new CreateFile() { Uri = codeBehindUri.ToString() })
+            };
+            
+            return new WorkspaceEdit()
+            {
+                Changes = changes,
+                DocumentChanges = documentChanges,
```

Comment:
I just spent some time reimplementing it to use TextDocumentChanges and it turns out that it doesn't really work as nicely due to how they are serialized (in comparison with CreateFile, RenameFile, and DeleteFile). Also, since we own the shim that translates these changes on the VSCode side, I think this is best left to be dealt with in the future (since as of right now it's not causing any issues).

Reply:
Sure, I mean you can actually remove setting `DocumentChanges` and only use `Changes`. It should still work as expected according to the spec.